app/Console/Commands/Takedate.php

```diff
@@ -29,80 +38,140 @@ class Takedate extends Command
+			$photoQuery->whereNull('taken_at');
+		}
+
+		$count = $photoQuery->count();
+		if ($count === 0) {
 			$this->line('No pictures require takedate updates.');
 
-			return false;
+			return -1;
 		}
 
-		$i = $from - 1;
+		$this->progressBar->setMaxSteps($count);
+		$photos = $photoQuery->lazyById();
 		/* @var Photo $photo */
 		foreach ($photos as $photo) {
-			$fullPath = $photo->full_path;
-			$i++;
+			$this->progressBar->advance();
+			// TODO: As soon as we support AWS S3 storage, we must stop using absolute paths. However, first the EXIF extractor must be rewritten to use file streams.
+			$fullPath = $photo->size_variants->getOriginal()->getFile()->getAbsolutePath();
+
 			if (!file_exists($fullPath)) {
-				$this->line($i . ': File ' . $fullPath . ' not found for ' . $photo->title . '.');
+				$this->printError('File ' . $fullPath . ' not found for photo "' . $photo->title . '" (ID=' . $photo->id . ').');
 				continue;
 			}
-			$info = $metadataExtractor->extract($fullPath, $photo->type);
-			/* @var \DateTime $stamp */
+
+			$kind = $photo->isRaw() ? 'raw' : ($photo->isVideo() ? 'video' : 'photo');
+			$info = $metadataExtractor->extract($fullPath, $kind);
+			/* @var Carbon $stamp */
 			$stamp = $info['taken_at'];
-			if ($stamp != null) {
-				if ($stamp == $photo->takestamp) {
-					$this->line($i . ': Takestamp up to date for ' . $photo->title);
+			if ($stamp !== null) {
+				// Note: `equalTo` only checks if two times indicate the same
+				// instant of time on the universe's timeline, i.e. equality
+				// comparison is always done in UTC.
+				// For example "2022-01-31 20:50 CET" is deemed equal to
+				// "2022-01-31 19:50 GMT".
+				// So, we must check for equality of timezones separately.
+				if ($photo->taken_at->equalTo($stamp) && $photo->taken_at->timezoneName === $stamp->timezoneName) {
+					$this->printInfo('Takestamp ' . $stamp->format(self::DATETIME_FORMAT) . ' up to date for photo "' . $photo->title . '" (ID=' . $photo->id . ').');
 					continue;
 				}
 				$photo->taken_at = $stamp;
 				if ($photo->save()) {
-					$this->line($i . ': Takestamp updated to ' . $stamp->format('d M Y \a\t H:i') . ' for ' . $photo->title);
+					$this->printInfo('Takestamp updated to ' . $stamp->format(self::DATETIME_FORMAT) . ' for photo "' . $photo->title . '" (ID=' . $photo->id . ').');
 				} else {
-					$this->line($i . ': Failed to update takestamp for ' . $photo->title);
+					$this->printError('Failed to update takestamp for photo "' . $photo->title . '" (ID=' . $photo->id . ').');
 				}
 				continue;
 			}
-			if (!$timestamps) {
-				$this->line($i . ': Failed to get Takestamp data for ' . $photo->title . '.');
+
+			// TODO: The description of this command's signature suggests something different than we do here.
+			// The description suggests that we would use the file date as a fall back for `taken_at`, if EXIF data is not available.
+			// However, we set `created_at` (aka upload time) here.
```

Comment:
That is surprising, although now that you mention it, it does seem to ring a bell. @ildyria, was it you who wanted it to work that way?

Reply:
Also keep in mind that we call `touch` on the media file during import and explicitly set the creation time of the file to the extracted `taken_at` attribute of the EXIF data. In this case, the reported creation time of the file does not match the upload time but indeed the takedate. However, if we cannot extract the EXIF information upon import, then we don't `touch` the media file and the creation date of the file remains equal to the upload time.

Considering that I am not sure if we can use the creation time of the file at all. We simply don't know if the creation time of the file has been faked to equal the extracted takedate or if the creation time of the file equals the import time :-(

_Question:_ Why do we `touch` the media file during import at all? If we did not do that we would at least know that the creation time of the file is always the upload time.

Also I am not sure if I like to  fiddle with the `created_at` attribute at all. Again, I am thinking about mobile clients which cash a shadow database for offline usage. The columns `created_at` and `updated_at` are used to calculate a 3-way merge if both the master and the shadow database were modified during an offline period. Fiddling around with these "internal", "technical" attributes might have consequences which I cannot asses, but it feels wrong to me.

Reply:
@ildyria Could you say something about this part? I would like to know, if I should do something about this before I merge the PR or not.

Reply:
the `touch` part is Legacy from Tobias:
https://github.com/LycheeOrg/Lychee-v3/blob/master/php/Modules/Photo.php#L256

Reply:
If nobody disagrees, I will do the following:

 * Drop the `touch` upon uploading. This way the timestamp of the file always consistently match the upload time. In particular, `touch` won't work anyway for Flysystem disks hosted on S3.
 * Allow to the set the upload time (aka `created_at`) from the timestamp of the filesystem as an _independent_ option, not as a fall-back if the takedate cannot be set. These two timestamps are completely independent.

Reply:
Applied changes as discussed.